mlflow/openai/__init__.py

```diff
@@ -0,0 +1,470 @@
+             metadata of the logged model.
+    """
+    return Model.log(
+        artifact_path=artifact_path,
+        flavor=mlflow.openai,
+        registered_model_name=registered_model_name,
+        model=model,
+        task=task,
+        conda_env=conda_env,
+        code_paths=code_paths,
+        signature=signature,
+        input_example=input_example,
+        await_registration_for=await_registration_for,
+        pip_requirements=pip_requirements,
+        extra_pip_requirements=extra_pip_requirements,
+        metadata=metadata,
+        **kwargs,
+    )
+
+
+def _load_model(path):
+    with open(path) as f:
+        return json.load(f)
+
+
+class _OpenAIWrapper:
+    def __init__(self, model):
+        self.model = model
+
+    def predict(self, dataframe):
+        import openai
+
+        model_dict = self.model.copy()
+        model_dict.pop("task", None)
+        model_dict.pop("messages", None)
+        responses = []
+        for messages in dataframe["messages"]:
+            completion = openai.ChatCompletion.create(
+                **model_dict,
+                # messages is a numpy array and not json serializable, hence `tolist()`
+                messages=messages.tolist(),
+            )
+            responses.append(completion.choices[0].message.content)
+        return responses
+
+
+_TEST_CONTENT = "test"
+
+
+@contextmanager
+def _mock_request_chat_completion():
+    response = requests.Response()
+    response.status_code = 200
+    response.headers = {"Content-Type": "application/json"}
+    # See https://platform.openai.com/docs/api-reference/chat/create for the response format
+    response._content = """
+{
+    "id": "chatcmpl-123",
+    "object": "chat.completion",
+    "created": 1677652288,
+    "choices": [{
+        "index": 0,
+        "message": {
+        "role": "assistant",
+        "content": "__TEST_CONTENT__"
+        },
+        "finish_reason": "stop"
+    }],
+    "usage": {
+        "prompt_tokens": 9,
+        "completion_tokens": 12,
+        "total_tokens": 21
+    }
+}
+    """.replace(
+        "__TEST_CONTENT__", _TEST_CONTENT
+    ).encode()
+    with mock.patch("requests.Session.request", return_value=response) as m:
+        yield m
+
+
+class _TestOpenAIWrapper(_OpenAIWrapper):
+    """
+    A wrapper class that should be used for testing purposes only.
+    """
+
+    def predict(self, dataframe):
+        with _mock_request_chat_completion() as m:
+            res = super().predict(dataframe)
+            m.assert_called_once()
+            return res
+
+
+def _load_pyfunc(path):
+    """
+    Load PyFunc implementation. Called by ``pyfunc.load_model``.
+
+    :param path: Local filesystem path to the MLflow Model with the ``openai`` flavor.
+    """
+    wrapper_cls = _TestOpenAIWrapper if _MLFLOW_OPENAI_TESTING.get() else _OpenAIWrapper
+    return wrapper_cls(_load_model(path))
```

Comment:
Could not find a good way to mock `requests` in the UDF

Reply:
Makes sense :)